### PR TITLE
Fix invalid timezone errors for valid timezones on Apple platforms.

### PR DESCRIPTION
**Pull request by chrsmys**
## Summary
This addresses issue #1607 where valid timezones were raising invalid time zone errors. The root of the issue is that hermes relies on `NSTimeZone.knownTimeZoneNames` to determine valid time zones, but this does not provide a complete list of all time zones NSTimeZone supports. `US/Eastern` is an example of a timezone that is not listed in knownTimeZoneNames but will generate a valid NSTimeZone.

To address this problem this will first attempt to create an NSTimeZone before raising the range exception.  If it's a valid time zone then it will get added to the validTimeZoneNames list, otherwise an exception will be raised. 

## Test Plan
To test this I built Hermes locally on a Mac and ran [this gist](https://gist.github.com/chrsmys/3af5462118798f2f090f11e226660c15) to detect which timezones do & do not work. I also tested negative cases to ensure exceptions were still raised when an invalid time zone is provided. Here is list of timezones that previously did not work, but now do:

```
Africa/Asmera
Africa/Timbuktu
America/Argentina/ComodRivadavia
America/Atka
America/Buenos_Aires
America/Catamarca
America/Coral_Harbour
America/Cordoba
America/Ensenada
America/Fort_Wayne
America/Indianapolis
America/Jujuy
America/Knox_IN
America/Louisville
America/Mendoza
America/Porto_Acre
America/Rosario
America/Virgin
Asia/Ashkhabad
Asia/Chungking
Asia/Dacca
Asia/Istanbul
Asia/Kolkata
Asia/Macao
Asia/Saigon
Asia/Tel_Aviv
Asia/Thimbu
Asia/Ujung_Pandang
Asia/Ulan_Bator
Atlantic/Faeroe
Atlantic/Jan_Mayen
Australia/ACT
Australia/Canberra
Australia/LHI
Australia/North
Australia/NSW
Australia/Queensland
Australia/South
Australia/Tasmania
Australia/Victoria
Australia/West
Australia/Yancowinna
Brazil/Acre
Brazil/DeNoronha
Brazil/East
Brazil/West
Canada/Atlantic
Canada/Central
Canada/Eastern
Canada/Mountain
Canada/Newfoundland
Canada/Pacific
Canada/Saskatchewan
Canada/Yukon
Chile/Continental
Chile/EasterIsland
CST6CDT
Cuba
Egypt
Eire
EST5EDT
Etc/GMT
Etc/GMT+0
Etc/GMT+1
Etc/GMT+10
Etc/GMT+11
Etc/GMT+12
Etc/GMT+2
Etc/GMT+3
Etc/GMT+4
Etc/GMT+5
Etc/GMT+6
Etc/GMT+7
Etc/GMT+8
Etc/GMT+9
Etc/GMT-0
Etc/GMT-1
Etc/GMT-10
Etc/GMT-11
Etc/GMT-12
Etc/GMT-13
Etc/GMT-14
Etc/GMT-2
Etc/GMT-3
Etc/GMT-4
Etc/GMT-5
Etc/GMT-6
Etc/GMT-7
Etc/GMT-8
Etc/GMT-9
Etc/GMT0
Etc/Greenwich
Etc/UCT
Etc/Universal
Etc/UTC
Etc/Zulu
Europe/Belfast
Europe/Nicosia
Europe/Tiraspol
Factory
GB
GB-Eire
GMT+0
GMT-0
GMT0
Greenwich
Hongkong
Iceland
Iran
Israel
Jamaica
Japan
Kwajalein
Libya
MET
Mexico/BajaNorte
Mexico/BajaSur
Mexico/General
MST7MDT
Navajo
NZ
NZ-CHAT
Pacific/Samoa
Pacific/Yap
Poland
Portugal
PRC
PST8PDT
ROC
ROK
Singapore
Turkey
UCT
Universal
US/Alaska
US/Aleutian
US/Arizona
US/Central
US/East-Indiana
US/Eastern
US/Hawaii
US/Indiana-Starke
US/Michigan
US/Mountain
US/Pacific
US/Samoa
W-SU
Zulu
```

### CODE REVIEW

**Comment by chrsmys** [comment]
@robhogan Do you know who might be best to take a look at this? Thanks!

**Comment by tienle46** [comment]
This PR should fix issue in our application. Please take a look soon.

**Comment by chrsmys** [comment]
@tmikov sorry for the direct ping, but do you know anyone who can take a look?

**Comment by tmikov** [comment]
Will do

File: lib/Platform/Intl/PlatformIntlApple.mm [hunk 1]
```diff
@@ -1462,8 +1462,12 @@ uint8_t getCurrencyDigits(std::u16string_view code) {
     timeZone = timeZoneIt->second.getString();
     // b. If the result of IsValidTimeZoneName(timeZone) is false, then
     if (!isValidTimeZoneName(timeZone)) {
-      // i. Throw a RangeError exception.
-      return runtime.raiseRangeError("Incorrect timeZone information provided");
+      if ([[NSTimeZone alloc] initWithName:u16StringToNSString(timeZone)] != nil) {
```
[the remark this answers is not in the document]
Comment (lavenzg) [reply lib/Platform/Intl/PlatformIntlApple.mm:1465]: And also, constructing NSTimeZone is not cheap, it'll be better to save the created timezone and pass it to initializeNSDateFormatter(), so that we don't need to create it again there.

File: lib/Platform/Intl/PlatformIntlApple.mm [hunk 2]
```diff
@@ -1462,8 +1462,12 @@ uint8_t getCurrencyDigits(std::u16string_view code) {
...
+          validTimeZoneNames().update(timeZone);
+      } else {
+          // i. Throw a RangeError exception.
+          return runtime.raiseRangeError("Incorrect timeZone information provided");
```
Comment (lavenzg) [inline lib/Platform/Intl/PlatformIntlApple.mm:1469]: nit: we didn't print out the actual timezone string in previous error message, let's add it now.

File: lib/Platform/Intl/PlatformIntlApple.mm (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrsmys) [reply lib/Platform/Intl/PlatformIntlApple.mm:1465]: @lavenzg None of the known time zones which are being used today work in there uppercased form, this feels like a separate issue that should be addressed outside of this PR:
```
for timeZone in TimeZone.knownTimeZoneIdentifiers {
    if NSTimeZone(name:timeZone.uppercased()) == nil {
        print(timeZone) <- Always nil
    }
}
```

> And also, constructing NSTimeZone is not cheap, it'll be better to save the created timezone and pass it to initializeNSDateFormatter(), so that we don't need to create it again there.

I can do this if we feel like that complexity is worth it. I am looking at this as more of a cache miss. There may be a small performance hit for missing the cache, but adding the time zone to the list of known time zones will make it so we only hit that issue once per app session. Thoughts?

File: lib/Platform/Intl/PlatformIntlApple.mm (hunk 2 above)
Comment (chrsmys) [reply lib/Platform/Intl/PlatformIntlApple.mm]: Done!

File: lib/Platform/Intl/PlatformIntlApple.mm (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrsmys) [reply lib/Platform/Intl/PlatformIntlApple.mm:1465]: > I tried in Swift that if the passed in string is "US/EASTERN", it returns nil. Both JSC/V8 can still recognize it.

Looking into this a bit and it looks like validTimeZoneNames stores the uppercased version of the time zone for lookup but uses the canonical time zone name when creating NSTimeZone. If that is true then Hermes should work the same way as JSC/V8. This change should just add more time zones that JSC & V8 also support.

[the remark this answers is not in the document]
Comment (lavenzg) [reply lib/Platform/Intl/PlatformIntlApple.mm:1465]: Oh yeah I remember it as the other direction. Then we can keep it.

> this feels like a separate issue that should be addressed outside of this PR

Sure, it's out of scope of this PR. 

> but adding the time zone to the list of known time zones will make it so we only hit that issue once per app session

There're cases handling a bunch of dates. But we can worry about that later. Maybe it doesn't matter comparing to the whole thing of constructing the DateFormatter.

**Review by lavenzg — APPROVED** [review]
LGTM.

**Comment by chrsmys** [comment]
> Cannot test target “ApplePlatformsIntegrationMobileTests” on “Any iOS Device”: Tests must be run on a concrete device

@lavenzg Looks like it's CI is failing to find a simulator to run on.

**Comment by chrsmys** [comment]
Looks like it might get fixed by this #1626

**Comment by lavenzg** [comment]
> > Cannot test target “ApplePlatformsIntegrationMobileTests” on “Any iOS Device”: Tests must be run on a concrete device
> 
> @lavenzg Looks like it's CI is failing to find a simulator to run on.

Yeah we are aware of that. Just ignore that for now.

**Comment by maxpowa** [comment]
Would love to see this fix backported into previous react-native versions of hermes - I had great success applying this patch to the v0.76-stable branch when running on iOS 17/18. Unfortunately, when running on iOS 15/16 the app immediately crashes with a `Symbol not found: __ZNSt3__122__libcpp_verbose_abortEPKcz` error. I don't think this error is specific to this patch, more just something wrong with my emulation of the react-native hermes framework build process.

**Comment by chrsmys** [comment]
@lavenzg just wanted to bump this in case it got lost

**Comment by lavenzg** [comment]
> @lavenzg just wanted to bump this in case it got lost

It has been merged. Thanks!

**Comment by navarroaxel** [comment]
@lavenzg could we please backport this to `rn/0.79-stable`? :pray:

**Comment by tmikov** [comment]
> @lavenzg could we please backport this to `rn/0.79-stable`? 🙏

Thanks for the request. We don't backport changes to older Hermes releases as a matter of policy. Our team focuses resources on forward development rather than maintaining multiple release branches.

Since Hermes is open source, you're welcome to create the backport yourself and submit a PR, if needed for your use case. Keep in mind that any backported changes would also need to be picked up by the corresponding React Native release branch to be useful.

**Comment by navarroaxel** [comment]
@tmikov Hello, I've created the PR to backport the commit. https://github.com/facebook/hermes/pull/1723
